src/main/java/com/streamr/client/utils/MessageCreationUtil.java

```diff
@@ -16,29 +17,69 @@
 import java.util.Map;
 
 public class MessageCreationUtil {
+
     private final String publisherId;
     private final String msgChainId;
     private final SigningUtil signingUtil;
+    private final Map<String, SecretKey> groupKeys = new HashMap<>(); // streamId --> groupKey
 
     private final HashMap<String, MessageRef> refsPerStreamAndPartition = new HashMap<>();
 
     private final HashMap<String, Integer> cachedHashes = new HashMap<>();
 
     public MessageCreationUtil(String publisherId, SigningUtil signingUtil) {
+        this(publisherId, signingUtil, new HashMap<>());
+    }
+
+    public MessageCreationUtil(String publisherId, SigningUtil signingUtil, Map<String, String> groupKeysHex) {
         this.publisherId = publisherId;
         msgChainId = RandomStringUtils.randomAlphanumeric(20);
         this.signingUtil = signingUtil;
+        for (String streamId: groupKeysHex.keySet()) {
+            String groupKeyHex = groupKeysHex.get(streamId);
+            EncryptionUtil.validateGroupKey(groupKeyHex);
+            groupKeys.put(streamId, new SecretKeySpec(DatatypeConverter.parseHexBinary(groupKeyHex), "AES"));
+        }
     }
 
     public StreamMessage createStreamMessage(Stream stream, Map<String, Object> payload, Date timestamp, String partitionKey) {
+        return createStreamMessage(stream, payload, timestamp, partitionKey, null);
+    }
+
+    public StreamMessage createStreamMessage(Stream stream, Map<String, Object> payload, Date timestamp, String partitionKey, String groupKeyHex) {
+        if (groupKeyHex != null) {
+            EncryptionUtil.validateGroupKey(groupKeyHex);
+        }
+
+        EncryptionType encryptionType;
+        String content;
+        if (groupKeys.containsKey(stream.getId()) && groupKeyHex != null) {
+            encryptionType = EncryptionType.NEW_KEY_AND_AES;
+            byte[] groupKeyBytes = DatatypeConverter.parseHexBinary(groupKeyHex);
+            byte[] payloadBytes = HttpUtils.mapAdapter.toJson(payload).getBytes(StandardCharsets.UTF_8);
```

Comment:
Relates to previous comment in `Subscription`: would be nice if this class knew nothing about the internals of how the encryption works (e.g. no assembling of group key and payload bytes), and no `HttpUtils.mapAdapter.toJson()` here